### PR TITLE
add DialURL function to simplify handling of ldap:// and ldaps:// urls

### PR DESCRIPTION
**Pull request by kofemann**
in some cases a client application wants simply to connect to an ldap
server based on provided ldap url. To simplify client-side code a
convenient function DialURL is added which based on provided url will
use plain tcp or tls connection. If no port number is provided, the
standard ports defined by IANA is used: 389 for plain and 636 for ssl
connections.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -126,6 +134,37 @@ func DialTLS(network, addr string, config *tls.Config) (*Conn, error) {
 	return conn, nil
 }
 
+// DialURL connects to the given ldap URL vie TCP using tls.Dial or net.Dial if ldaps://
+// or ldap:// specified as protocol. On success a new Conn for the connection
+// is returned.
+func DialURL(addr string) (*Conn, error) {
```
Comment (vetinari) [inline conn.go]: can we have an *tls.Config as second parameter which is used when non-nil for TLS connections?

File: conn.go [hunk 2]
```diff
@@ -126,6 +134,37 @@ func DialTLS(network, addr string, config *tls.Config) (*Conn, error) {
...
+
+	lurl, err := url.Parse(addr)
+	if err != nil {
+		return nil, NewError(ErrorNetwork, err)
+	}
+
```
[the remark this answers is not in the document]
Comment (kofemann) [reply conn.go]: Ok. makes sense. However, I cant use net.SplitHostPort directly as it returns an error it port is not provided.

File: conn.go (hunk 1 above)
Comment (kofemann) [reply conn.go]: Someone why wants full control should use DialTLS. This function is intended to be a short-cut for common usage.

File: conn.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply conn.go]: taking the error to mean the port wasn't provided (or wasn't valid) is typical

also, use JoinHostPort rather than concatenating with ":" (for ipv6 safety)

[the remark this answers is not in the document]
Comment (kofemann) [reply conn.go]: @liggitt it's already done with latest update

File: conn.go [hunk 3]
```diff
@@ -457,3 +497,18 @@ func (l *Conn) reader() {
...
+// a variation of net.SplitHostPort which returns string as port if not specified.
+func splitHostAndPort(hostport string) (string, string, error) {
+
+	if hostport == "" {
+		return "", "", errors.New("Empty hostport provided")
+	}
+
+	// make golang < 1.8 compatible
+	i := strings.LastIndexByte(hostport, ':')
```
Comment (liggitt) [inline conn.go]: this fails on go 1.4: https://travis-ci.org/go-ldap/ldap/jobs/433990177

File: conn.go [hunk 4]
```diff
@@ -457,3 +497,18 @@ func (l *Conn) reader() {
...
+	if i == -1 || i < strings.IndexByte(hostport, ']') {
```
Comment (liggitt) [inline conn.go]: this doesn't look proper... why is this needed?

Comment (kofemann) [reply conn.go]: this test checks that collon is after braces, e,g, to exclude IPv6 address without port, like **[::1]**

File: conn.go (hunk 3 above)
Comment (kofemann) [reply conn.go]: fixed. The build for 1.9  failed, but I believe it's not related to my changes.

File: conn.go (hunk 4 above)
Comment (liggitt) [reply conn.go]: I think we should simply use SplitHostPort and assume the input is either a valid host or host:port

Comment (kofemann) [reply conn.go]: SplitHostPort returns an error when port is not specified. This makes it hard to identify invalid string from missing port, unless we want to parse the error.

Comment (liggitt) [reply conn.go]: we shouldn't try to identify an invalid string... if url.Parse passes, we should accept the host blob. if SplitHostPort fails, we should assume it is a port-less host

Comment (kofemann) [reply conn.go]: probably will cover the majority of usecases.... :+1:

**Comment by kofemann** [comment]
ping

**Comment by johnweldon** [comment]
@kofemann Thanks for the updates - can you update your branch with the latest on master and we'll see about getting this merged.

**Comment by kofemann** [comment]
@johnweldon rebased on top of current master (5261598)
